File: blocks/block_03_dsp_basics/cpp/lab3_digital_mixing.cpp
#include <cmath>
#include <complex>
#include <fstream>
#include <iostream>
#include <vector>

int main() {
    constexpr double Fs = 1.0e6;
    constexpr int N = 4096;
    constexpr double f_sig = 50.0e3;
    constexpr double f_shift = 100.0e3;
    constexpr double pi = 3.14159265358979323846;

    std::vector<std::complex<double>> x(N);
    std::vector<std::complex<double>> y(N);

    for (int n = 0; n < N; ++n) {
        const double t = static_cast<double>(n) / Fs;
        const std::complex<double> sig = std::exp(std::complex<double>(0.0, 2.0 * pi * f_sig * t));
        const std::complex<double> nco = std::exp(std::complex<double>(0.0, 2.0 * pi * f_shift * t));
        x[n] = sig;
        y[n] = sig * nco;
    }

    std::ofstream out("lab3_digital_mixing_iq.csv");
    out << "n,x_i,x_q,y_i,y_q\n";
    for (int n = 0; n < N; ++n) {
        out << n << "," << x[n].real() << "," << x[n].imag()
            << "," << y[n].real() << "," << y[n].imag() << "\n";
    }

    std::cout << "Generated lab3_digital_mixing_iq.csv\n";
    std::cout << "Expected frequency shift: " << f_shift << " Hz\n";
    std::cout << "Expected output tone: " << (f_sig + f_shift) << " Hz\n";

    return 0;
}

File: blocks/block_03_dsp_basics/python/lab3_digital_mixing.py
import numpy as np
import matplotlib.pyplot as plt

Fs = 1e6
N = 4096

t = np.arange(N)/Fs

# input tone
f_sig = 50e3
x = np.exp(1j*2*np.pi*f_sig*t)

# frequency shift
f_shift = 100e3
nco = np.exp(1j*2*np.pi*f_shift*t)

y = x * nco

# FFT helper

def plot_fft(sig, title):
    X = np.fft.fftshift(np.fft.fft(sig))
    f = np.fft.fftshift(np.fft.fftfreq(len(sig), 1/Fs))
    plt.plot(f/1e3, 20*np.log10(np.abs(X)+1e-12))
    plt.title(title)
    plt.xlabel("Frequency (kHz)")
    plt.ylabel("Magnitude (dB)")

plt.figure(figsize=(10,6))

plt.subplot(2,1,1)
plot_fft(x, "Before mixing")

plt.subplot(2,1,2)
plot_fft(y, "After mixing")

plt.tight_layout()
plt.show()
